x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.type;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.esql.parser.ExpressionBuilder;
+import org.elasticsearch.xpack.esql.parser.ParsingException;
+import org.elasticsearch.xpack.ql.QlIllegalArgumentException;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.Converter;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypeConverter;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Period;
+import java.time.temporal.TemporalAmount;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.ql.type.DataTypes.NULL;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isPrimitive;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isString;
+
+public class EsqlDataTypeConverter {
+
+    /**
+     * Returns true if the from type can be converted to the to type, false - otherwise
+     */
+    public static boolean canConvert(DataType from, DataType to) {
+        // Special handling for nulls and if conversion is not requires
+        if (from == to || from == NULL) {
+            return true;
+        }
+        // only primitives are supported so far
+        return isPrimitive(from) && isPrimitive(to) && converterFor(from, to) != null;
+    }
+
+    public static Converter converterFor(DataType from, DataType to) {
+        Converter converter = DataTypeConverter.converterFor(from, to);
+        if (converter != null) {
+            return converter;
+        }
+        if (isString(from) && to == EsqlDataTypes.TIME_DURATION) {
+            return EsqlConverter.STRING_TO_TIME_DURATION;
+        }
+        if (isString(from) && to == EsqlDataTypes.DATE_PERIOD) {
+            return EsqlConverter.STRING_TO_DATE_PERIOD;
+        }
+        return null;
+    }
+
+    public static TemporalAmount parseTemporalAmount(Object val, DataType expectedType) {
+        String str = String.valueOf(val);
+        if (str == null) {
+            return null;
+        }
+        str = str.trim().replaceAll("\\s+", " ");
+        String[] split = str.split(" ");
```

Comment:
Replace the regext with a more hand-on/low-level approach just to be safe as some generated queries might be large yet due to typos might get picked here.
The advantage is that if the more than 2 entries are found, the code can bail out instead of parsing the whole string.